File: Implementations/my_array.py
#Implementing an array

class Array():
    def __init__(self):
        self.length = 0
        self.data = {}
    
    def get(self, index): #O(1)
        return self.data[index]    
    
    def append(self, item): #O(1)
        self.data[self.length] = item
        self.length += 1
        return self
    
    def pop(self): #O(1)
        lastItem = self.data[self.length-1]
        del self.data[self.length-1]
        self.length -= 1
        return lastItem
    
    def delete(self, index): #O(n)
        item = self.data[index]
        self.shift(index)
        
    def shift(self, index): #O(n)
        for i in range(index, self.length -1):
            self.data[i] = self.data[i+1]
        del self.data[self.length-1]
        self.length -= 1
    


    

File: cdsa/my_array.c
#include "my_array.h"
#include <stdlib.h>

static int  _get    (Array *self, int index)  { return self->data[index]; }
static int  _pop    (Array *self)             { int v = self->data[--self->length]; return v; }
static void _destroy(Array *self)             { free(self->data); free(self); }

static void _append(Array *self, int item) {
    if (self->length == self->capacity) {
        self->capacity *= 2;
        self->data = (int *)realloc(self->data, self->capacity * sizeof(int));
    }
    self->data[self->length++] = item;
}

static void _shift(Array *self, int index) {
    for (int i = index; i < self->length - 1; i++)
        self->data[i] = self->data[i + 1];
    self->length--;
}

static void _delete(Array *self, int index) { _shift(self, index); }

Array *Array_new() {
    Array *a    = (Array *)malloc(sizeof(Array));
    a->length   = 0;
    a->capacity = 4;
    a->data     = (int *)malloc(a->capacity * sizeof(int));
    a->get      = _get;
    a->append   = _append;
    a->pop      = _pop;
    a->delete   = _delete;
    a->destroy  = _destroy;
    return a;
}
